### PR TITLE
Fix PyTorch Hub export inference shapes

### LINKED ISSUE #6947: After converting from .pt to .torchscript, bounding box is off
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am playing around with YOLOv5 and exporting to different formats (.pt & .torchscript). But I found that after I exported the pre-trained models using this command: 
`!python "/content/drive/MyDrive/yolov5/export.py" --weights "/content/drive/MyDrive/yolov5/yolov5s.pt" --include torchscript`

The .torchscript model's bounding is off, please advise me on why this is happening and how to resolve it. Thank you.

**.pt model:**
zidane1

**.torchscript model:**
zidane2

### Additional

_No response_

### PR DESCRIPTION
**Pull request by glenn-jocher**
May resolve https://github.com/ultralytics/yolov5/issues/6947

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Simplified image preprocessing in the YOLOv5 forward pass method.

### 📊 Key Changes
- Removed conditional logic for different inference shapes based on `self.pt`.
- Streamlined the process for making inference shape divisible by the stride and applying letterbox padding to images.
- Improved image stacking and format conversion (BHWC to BCHW) process.

### 🎯 Purpose & Impact
- Enhances code readability and maintainability by reducing complexity.
- Ensures consistent image preprocessing regardless of the PyTorch (`self.pt`) presence.
- Potentially increases efficiency during the image preprocessing step, leading to faster image handling in the forward pass.

These changes can provide a smoother and more understandable experience for developers working with the YOLOv5 codebase, while end-users might benefit from slight performance improvements during model inference. 🚀

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Validation results:
```python
!python export.py --weights yolov5s.pt --include torchscript
model = torch.hub.load('ultralytics/yolov5', 'custom', path='yolov5s.torchscript')
result = model("data/images/zidane.jpg")
result.save()
```
Screenshot 2022-03-11 at 16 16 14

**Comment by mynameischaos** [comment]
@glenn-jocher  Hi, I run the same scripts.
python export.py --weights yolov5s.pt --include torchscript --device 0
Then, run:

import torch
model = torch.hub.load('ultralytics/yolov5', 'custom', path='yolov5s.pt')
result = model("data/images/zidane.jpg")
result.save()
zidane_pt



import torch
model = torch.hub.load('ultralytics/yolov5', 'custom', path='yolov5s.torchscript')
result = model("data/images/zidane.jpg")
result.save()
zidane_torchscript

But the results are different. (the score of bounding box.)

**Comment by glenn-jocher** [comment]
@mynameischaos all export formats are benchmarked every 24 hours and on every commit here to verify equal mAP. Last run 14 hours ago:
https://github.com/ultralytics/yolov5/runs/7531416890?check_suite_focus=true
Screen Shot 2022-07-27 at 4 30 05 PM

**Comment by mynameischaos** [comment]
@glenn-jocher  

I use this following script to convet model to torchscript format:

python3 export.py --weights yolov5n.pt  --img 640 --batch 1 --device 0

Then, I run the validation scripts:

python val.py --weights yolov5n.pt     ->   result
P          R     mAP@.5 mAP@.5:.95
0.612      0.492      0.549      0.349

python val.py --weights yolov5n.torchscript       ->   result
P          R     mAP@.5 mAP@.5:.95
0.647      0.462      0.545      0.348

I cannot get the same results. Why?

**Comment by glenn-jocher** [comment]
@mynameischaos 👋 Hello! Thanks for asking about YOLOv5 🚀 **benchmarks**. YOLOv5 inference is officially supported in 11 formats, and all formats are benchmarked for identical accuracy and to compare speed every 24 hours by the [YOLOv5 CI](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml?query=event%3Aschedule).

💡 ProTip: Export to ONNX or OpenVINO for up to 3x CPU speedup. See [CPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6613).
💡 ProTip: Export to TensorRT for up to 5x GPU speedup. See [GPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6963).

Format                      | `export.py --include`         | Model
:---                        | --:                           | :--
[PyTorch](https://pytorch.org/)                     | -                             | `yolov5s.pt`
[TorchScript](https://pytorch.org/docs/stable/jit.html)                 | `torchscript`                 | `yolov5s.torchscript`
[ONNX](https://onnx.ai/)                        | `onnx`                        | `yolov5s.onnx`
[OpenVINO](https://docs.openvino.ai/latest/index.html)                    | `openvino`                    | `yolov5s_openvino_model/`
[TensorRT](https://developer.nvidia.com/tensorrt)                    | `engine`                      | `yolov5s.engine`
[CoreML](https://github.com/apple/coremltools)                      | `coreml`                      | `yolov5s.mlmodel`
[TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model)       | `saved_model`                 | `yolov5s_saved_model/`
[TensorFlow GraphDef](https://www.tensorflow.org/api_docs/python/tf/Graph)         | `pb`                          | `yolov5s.pb`
[TensorFlow Lite](https://www.tensorflow.org/lite)             | `tflite`                      | `yolov5s.tflite`
[TensorFlow Edge TPU](https://coral.ai/docs/edgetpu/models-intro/)         | `edgetpu`                     | `yolov5s_edgetpu.tflite`
[TensorFlow.js](https://www.tensorflow.org/js)               | `tfjs`                        | `yolov5s_web_model/`


## Benchmarks

Benchmarks below run on a Colab Pro with the YOLOv5 tutorial notebook <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb">Open In Colab</a>. To reproduce:
```bash
python utils/benchmarks.py --weights yolov5s.pt --imgsz 640 --device 0
```

### Colab Pro V100 GPU

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=0, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 46.7/166.8 GB disk)

Benchmarks complete (458.07s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623                10.19
1             TorchScript        0.4623                 6.85
2                    ONNX        0.4623                14.63
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4617                 1.89
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623                21.28
7     TensorFlow GraphDef        0.4623                21.22
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

### Colab Pro CPU

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=cpu, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CPU
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 41.5/166.8 GB disk)

Benchmarks complete (241.20s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623               127.61
1             TorchScript        0.4623               131.23
2                    ONNX        0.4623                69.34
3                OpenVINO        0.4623                66.52
4                TensorRT           NaN                  NaN
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623               123.79
7     TensorFlow GraphDef        0.4623               121.57
8         TensorFlow Lite        0.4623               316.61
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

Good luck 🍀 and let us know if you have any other questions!

**Comment by mynameischaos** [comment]
> @glenn-jocher
> 
> I use this following script to convet model to torchscript format:
> 
> python3 export.py --weights yolov5n.pt --img 640 --batch 1 --device 0
> 
> Then, I run the validation scripts:
> 
> python val.py --weights yolov5n.pt -> result P R mAP@.5 mAP@.5:.95 0.612 0.492 0.549 0.349
> 
> python val.py --weights yolov5n.torchscript -> result P R mAP@.5 mAP@.5:.95 0.647 0.462 0.545 0.348
> 
> I cannot get the same results. Why?

Please help me about this problem. @glenn-jocher

**Comment by glenn-jocher** [comment]
@mynameischaos 👋 Hello! Thanks for asking about YOLOv5 🚀 **benchmarks**. YOLOv5 inference is officially supported in 11 formats, and all formats are benchmarked for identical accuracy and to compare speed every 24 hours by the [YOLOv5 CI](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml?query=event%3Aschedule).

💡 ProTip: Export to ONNX or OpenVINO for up to 3x CPU speedup. See [CPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6613).
💡 ProTip: Export to TensorRT for up to 5x GPU speedup. See [GPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6963).

Format                      | `export.py --include`         | Model
:---                        | --:                           | :--
[PyTorch](https://pytorch.org/)                     | -                             | `yolov5s.pt`
[TorchScript](https://pytorch.org/docs/stable/jit.html)                 | `torchscript`                 | `yolov5s.torchscript`
[ONNX](https://onnx.ai/)                        | `onnx`                        | `yolov5s.onnx`
[OpenVINO](https://docs.openvino.ai/latest/index.html)                    | `openvino`                    | `yolov5s_openvino_model/`
[TensorRT](https://developer.nvidia.com/tensorrt)                    | `engine`                      | `yolov5s.engine`
[CoreML](https://github.com/apple/coremltools)                      | `coreml`                      | `yolov5s.mlmodel`
[TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model)       | `saved_model`                 | `yolov5s_saved_model/`
[TensorFlow GraphDef](https://www.tensorflow.org/api_docs/python/tf/Graph)         | `pb`                          | `yolov5s.pb`
[TensorFlow Lite](https://www.tensorflow.org/lite)             | `tflite`                      | `yolov5s.tflite`
[TensorFlow Edge TPU](https://coral.ai/docs/edgetpu/models-intro/)         | `edgetpu`                     | `yolov5s_edgetpu.tflite`
[TensorFlow.js](https://www.tensorflow.org/js)               | `tfjs`                        | `yolov5s_web_model/`


## Benchmarks

Benchmarks below run on a Colab Pro with the YOLOv5 tutorial notebook <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb">Open In Colab</a>. To reproduce:
```bash
python utils/benchmarks.py --weights yolov5s.pt --imgsz 640 --device 0
```

### Colab Pro V100 GPU

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=0, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 46.7/166.8 GB disk)

Benchmarks complete (458.07s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623                10.19
1             TorchScript        0.4623                 6.85
2                    ONNX        0.4623                14.63
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4617                 1.89
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623                21.28
7     TensorFlow GraphDef        0.4623                21.22
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

### Colab Pro CPU

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=cpu, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CPU
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 41.5/166.8 GB disk)

Benchmarks complete (241.20s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623               127.61
1             TorchScript        0.4623               131.23
2                    ONNX        0.4623                69.34
3                OpenVINO        0.4623                66.52
4                TensorRT           NaN                  NaN
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623               123.79
7     TensorFlow GraphDef        0.4623               121.57
8         TensorFlow Lite        0.4623               316.61
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

Good luck 🍀 and let us know if you have any other questions!
